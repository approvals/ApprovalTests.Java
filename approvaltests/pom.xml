<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">

    <modelVersion>4.0.0</modelVersion>

    <parent>
        <artifactId>approvaltests-parent</artifactId>
        <groupId>com.approvaltests</groupId>
        <version>100.0.0-SNAPSHOT</version>
    </parent>

    <artifactId>approvaltests</artifactId>
    <packaging>jar</packaging>


    <dependencies>
        <dependency>
            <groupId>com.approvaltests</groupId>
            <artifactId>approvaltests-util</artifactId>
            <version>${project.parent.version}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.camel</groupId>
            <artifactId>camel-xstream</artifactId>
            <version>3.22.2</version>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>org.testng</groupId>
            <artifactId>testng</artifactId>
            <version>7.5.1</version>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>com.thoughtworks.xstream</groupId>
            <artifactId>xstream</artifactId>
            <version>1.4.20</version>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>org.codehaus.jettison</groupId>
            <artifactId>jettison</artifactId>
            <version>1.5.4</version>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>com.google.code.gson</groupId>
            <artifactId>gson</artifactId>
            <version>2.10.1</version>
            <scope>compile</scope>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-databind</artifactId>
            <version>2.17.2</version>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>commons-net</groupId>
            <artifactId>commons-net</artifactId>
            <version>3.11.1</version>
            <scope>compile</scope>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>quartz</groupId>
            <artifactId>quartz</artifactId>
            <version>1.5.2</version>
            <scope>compile</scope>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>commons-httpclient</groupId>
            <artifactId>commons-httpclient</artifactId>
            <version>3.1</version>
            <scope>compile</scope>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>jakarta.servlet</groupId>
            <artifactId>jakarta.servlet-api</artifactId>
            <version>5.0.0</version>
            <optional>true</optional>
        </dependency>
        <dependency>
                <groupId>org.apache.velocity</groupId>
                <artifactId>velocity-engine-core</artifactId>
             <version>2.3</version>
            <scope>compile</scope>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>junit</groupId>
            <artifactId>junit</artifactId>
            <version>4.13.2</version>
            <scope>compile</scope>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-api</artifactId>
            <version>5.10.3</version>
            <scope>compile</scope>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter</artifactId>
            <version>5.10.2</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-engine</artifactId>
            <version>5.10.3</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.junit.vintage</groupId>
            <artifactId>junit-vintage-engine</artifactId>
            <version>5.10.3</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>com.github.testng-team</groupId>
            <artifactId>testng-junit5</artifactId>
            <version>0.0.1</version>
            <scope>test</scope>
        </dependency>

    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-jar-plugin</artifactId>
                <version>3.4.2</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>test-jar</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <artifactId>maven-assembly-plugin</artifactId>
                <version>3.7.1</version>
                <configuration>
                    <descriptors>
                        <descriptor>src/main/assembly/bin.xml</descriptor>
                    </descriptors>
                    <appendAssemblyId>false</appendAssemblyId>
                </configuration>
                <executions>
                    <execution>
                        <id>make-assembly</id> <!-- this is used for inheritance merges -->
                        <phase>package</phase> <!-- append to the packaging phase. -->
                        <goals>
                            <goal>single</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>

</project>
