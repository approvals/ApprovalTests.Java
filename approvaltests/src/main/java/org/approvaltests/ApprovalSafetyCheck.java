package org.approvaltests;

import org.approvaltests.core.Options;
import org.lambda.actions.Action1;

public class ApprovalSafetyCheck {
    public static final ThreadLocal<Action1<Options>> guardRailCheck = ThreadLocal.withInitial(Action1::doNothing);

    public static void checkGuardRails(Options options) {
        Action1<Options> check = guardRailCheck.get();
        guardRailCheck.set(Action1.doNothing());
        check.call(options);
    }

    public static void setGuardRailCheck(Action1<Options> check) {
        guardRailCheck.set(check);
    }
}