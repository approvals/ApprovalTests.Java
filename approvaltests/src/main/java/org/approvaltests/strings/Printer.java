package org.approvaltests.strings;

/**
 * @deprecated use lambdas: {@code  a  -> from + " => " + to + "\n" }
 */
@Deprecated
public class Printer<T>
{
  public static String multiline(String from, String to)
  {
    return (from + "\n => \n" + to + "\n" + "------------------------------------");
  }

  public static String printLine(Object from, Object to)
  {
    return from + " => " + to + "\n";
  }
}
