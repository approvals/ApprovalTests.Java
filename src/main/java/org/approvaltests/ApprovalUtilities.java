package org.approvaltests;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

public class ApprovalUtilities
{
  public ByteArrayOutputStream writeSystemOutToStringBuffer()
  {
    ByteArrayOutputStream text = new ByteArrayOutputStream();
    System.setOut(new PrintStream(text));
    return text;
  }
}
