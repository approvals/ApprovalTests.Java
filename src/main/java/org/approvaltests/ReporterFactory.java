package org.approvaltests;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.FileLauncherReporter;
import org.approvaltests.reporters.ImageReporter;
import org.approvaltests.reporters.MultiReporter;
import org.approvaltests.reporters.QuietReporter;
import org.approvaltests.reporters.UseReporter;

import com.spun.util.ClassUtils;

public class ReporterFactory
{
  private static HashMap<String, Class<? extends ApprovalFailureReporter>> reporters = new HashMap<String, Class<? extends ApprovalFailureReporter>>();
  public static class FileTypes
  {
    public static final String  Text    = "txt";
    public static final String  Html    = "html";
    public static final String  Excel   = "csv";
    public static final String  File    = "file";
    public static final String  Image   = "png";
    private static final String Default = "default";
  }
  static
  {
    setupReporters();
  }
  public static ApprovalFailureReporter get(String string)
  {
    ApprovalFailureReporter returned = getFromAnnotation();
    returned = tryFor(returned, reporters.get(string));
    returned = tryFor(returned, reporters.get(FileTypes.Default));
    return returned;
  }
  public static ApprovalFailureReporter getFromAnnotation()
  {
    UseReporter reporter = getAnnotationFromStackTrace(UseReporter.class);
    return reporter == null ? null : getReporter(reporter);
  }
  private static ApprovalFailureReporter getReporter(UseReporter reporter)
  {
    Class<? extends ApprovalFailureReporter>[] classes = reporter.value();
    List<ApprovalFailureReporter> reporters = new ArrayList<ApprovalFailureReporter>();
    for (Class<? extends ApprovalFailureReporter> clazz : classes)
    {
      ApprovalFailureReporter instance = ClassUtils.create(clazz);
      reporters.add(instance);
    }
    return reporters.size() == 1 ? reporters.get(0) : new MultiReporter(reporters);
  }
  private static <T extends Annotation> T getAnnotationFromStackTrace(Class<T> annotationClass)
  {
    StackTraceElement[] trace = Thread.currentThread().getStackTrace();
    for (StackTraceElement stack : trace)
    {
      Method method = null;
      Class<?> clazz = null;
      try
      {
        String methodName = stack.getMethodName();
        clazz = Class.forName(stack.getClassName());
        method = clazz.getMethod(methodName, null);
      }
      catch (Exception e)
      {
        //ignore
      }
      T annotation = null;
      if (method != null)
      {
        annotation = method.getAnnotation(annotationClass);
      }
      if (annotation != null) { return annotation; }
      annotation = clazz.getAnnotation(annotationClass);
      if (annotation != null) { return annotation; }
    }
    return null;
  }
  private static ApprovalFailureReporter tryFor(ApprovalFailureReporter returned,
      Class<? extends ApprovalFailureReporter> trying)
  {
    if (returned == null && trying != null) { return ClassUtils.create(trying); }
    return returned;
  }
  private static void setupReporters()
  {
    reporters.put(FileTypes.Text, DiffReporter.class);
    reporters.put(FileTypes.Html, DiffReporter.class);
    reporters.put(FileTypes.Excel, FileLauncherReporter.class);
    reporters.put(FileTypes.File, FileLauncherReporter.class);
    reporters.put(FileTypes.Image, ImageReporter.class);
    reporters.put(FileTypes.Default, QuietReporter.class);
  }
  public static void clearAllReportersExceptDefault()
  {
    Class all = reporters.get(FileTypes.Default);
    reporters.clear();
    reporters.put(FileTypes.Default, all);
  }
}
