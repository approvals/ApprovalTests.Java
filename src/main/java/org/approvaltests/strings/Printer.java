package org.approvaltests.strings;

import org.lambda.functions.implementations.F1;

public class Printer<T> extends F1<T, String>
{
  public Printer(T a, Object... extraVariables)
  {
    super(a, extraVariables);
  }
  public void multiline(String from, String to)
  {
    ret(from + "\r\n => \r\n" + to + "\r\n" + "------------------------------------");
  }
  public static String printLine(Object from, Object to)
  {
    return from + " => " + to + "\r\n";
  }
  public void print(Object from, Object to)
  {
    ret(from + " => " + to);
  }
  public void format(String format, Object... params)
  {
    ret(String.format(format, params));
  }
}
