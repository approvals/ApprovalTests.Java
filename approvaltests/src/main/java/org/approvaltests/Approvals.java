package org.approvaltests;

import com.spun.util.ArrayUtils;
import com.spun.util.FormattedException;
import com.spun.util.JsonUtils;
import com.spun.util.ObjectUtils;
import com.spun.util.StringUtils;
import com.spun.util.persistence.ExecutableCommand;
import com.spun.util.persistence.Loader;
import com.spun.util.persistence.SqlLoader;
import org.approvaltests.approvers.ApprovalApprover;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.*;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.MasterDirectoryNamer;
import org.approvaltests.namer.NamerFactoryForOptions;
import org.approvaltests.namer.StackTraceNamer;
import org.approvaltests.reporters.ExecutableQueryFailure;
import org.approvaltests.writers.ApprovalXmlWriter;
import org.lambda.actions.Action0;
import org.lambda.functions.Function1;
import org.lambda.query.Query;

import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class Approvals
{
  public static final NamerFactoryForOptions NAMES        = NamerFactoryForOptions.INSTANCE;
  public static Loader<ApprovalNamer>        namerCreater = new Loader<ApprovalNamer>()
                                                          {
                                                            public ApprovalNamer load()
                                                            {
                                                              return new StackTraceNamer();
                                                            }
                                                          };
  public static void verify(String response)
  {
    verify(response, new Options());
  }
  public static void verify(String response, Options options)
  {
    verify(options.createWriter(response), options);
  }
  public static void verify(Object object)
  {
    verify(object, new Options());
  }
  public static void verify(Object object, Options options)
  {
    checkForAwtComponents(object == null ? Object.class : object.getClass());
    verify(Objects.toString(object), options);
  }
  public static void verify(Verifiable object)
  {
    verify(object, new Options());
  }
  public static void verify(Verifiable object, Options options)
  {
    VerifyParameters verifyParameters = object.getVerifyParameters(options);
    verify(Objects.toString(object), verifyParameters.getOptions());
  }
  private static void checkForAwtComponents(Class<?> type)
  {
    if (type == Object.class)
    { return; }
    if ("java.awt.Component".equals(type.getCanonicalName()))
    {
      throw new FormattedException(
          "Approvals.verify(Component) has moved to AwtApprovals.verify(Component)\ncalled Approvals.verify(Object)");
    }
    checkForAwtComponents(type.getSuperclass());
  }
  public static <T> void verifyAll(String label, T[] array)
  {
    verifyAll(label, array, new Options());
  }
  public static <T> void verifyAll(String label, T[] array, Options options)
  {
    verify(StringUtils.toString(label, array), options);
  }
  public static <T> void verifyAll(String header, String label, T[] array)
  {
    verifyAll(header, label, array, new Options());
  }
  public static <T> void verifyAll(String header, String label, T[] array, Options options)
  {
    String text = formatHeader(header) + StringUtils.toString(label, array);
    verify(text, options);
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1)
  {
    verifyAll(values, f1, new Options());
  }
  public static <T> void verifyAll(Iterable<T> values, Function1<T, String> f1)
  {
    verifyAll(values, f1, new Options());
  }
  public static <T> void verifyAll(Iterable<T> values, Function1<T, String> f1, Options options)
  {
    verifyAll("", values, f1, options);
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1, Options options)
  {
    String text = ArrayUtils.toString(values, f1);
    verify(text, options);
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, values, f1, new Options());
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1, Options options)
  {
    verifyAll(header, Arrays.asList(values), f1, options);
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1)
  {
    verifyAll(header, array, f1, new Options());
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1, Options options)
  {
    String text = formatHeader(header) + ArrayUtils.toString(array, f1);
    verify(text, options);
  }
  private static String formatHeader(String header)
  {
    return StringUtils.isEmpty(header) ? "" : header + "\n\n\n";
  }
  public static <T> void verifyAll(String label, Iterable<T> array)
  {
    verifyAll(label, array, new Options());
  }
  public static <T> void verifyAll(String label, Iterable<T> array, Options options)
  {
    verify(StringUtils.toString(label, array), options);
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array)
  {
    verifyAll(header, label, array, new Options());
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array, Options options)
  {
    String text = formatHeader(header) + StringUtils.toString(label, array);
    verify(text, options);
  }
  public static void verifyHtml(String response)
  {
    verifyHtml(response, new Options());
  }
  public static void verifyHtml(String response, Options options)
  {
    verify(response, options.forFile().withExtension(".html"));
  }
  public static void verify(File generateFile)
  {
    verify(generateFile, new Options());
  }
  public static void verify(File generateFile, Options options)
  {
    verify(options.createWriter(generateFile), options);
  }
  /**
   * @deprecated Use {@link #verify(ApprovalWriter, ApprovalNamer, Options)} instead.
   */
  @Deprecated
  public static void verify(ApprovalWriter writer, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(new FileApprover(writer, namer), new Options(reporter));
  }
  public static void verify(ApprovalWriter writer, ApprovalNamer namer)
  {
    verify(writer, namer, new Options());
  }
  public static void verify(ApprovalWriter writer, ApprovalNamer namer, Options options)
  {
    verify(new FileApprover(writer, namer, options.getComparator()), options);
  }
  public static void verify(ApprovalWriter writer)
  {
    verify(writer, new Options());
  }
  public static void verify(ApprovalWriter writer, Options options)
  {
    verify(writer, options.forFile().getNamer(), options);
  }
  public static void verifyXml(String xml)
  {
    verifyXml(xml, new Options());
  }
  public static void verifyXml(String xml, Options options)
  {
    final String formattedXml = ApprovalXmlWriter.prettyPrint(xml, 2);
    verify(formattedXml, options.forFile().withExtension(".xml"));
  }
  public static void verify(ApprovalApprover approver)
  {
    verify(approver, new Options());
  }
  /**
   * @deprecated Use {@link #verify(ApprovalApprover, Options)} instead.
   */
  @Deprecated
  public static void verify(ApprovalApprover approver, ApprovalFailureReporter reporter)
  {
    verify(approver, new Options(reporter));
  }
  public static void verify(ApprovalApprover approver, Options options)
  {
    ApprovalFailureReporter reporter = options.getReporter();
    VerifyResult result = approver.approve();
    if (result.isFailure())
    {
      result = approver.reportFailure(reporter);
    }
    if (result.isSuccessful())
    {
      approver.cleanUpAfterSuccess(reporter);
    }
    else
    {
      approver.fail();
    }
  }
  public static void verify(ExecutableCommand query)
  {
    verify(query, new Options());
  }
  public static void verify(ExecutableCommand query, Options options)
  {
    verify(query.getCommand(), ExecutableQueryFailure.create(query, options));
  }
  public static void verify(Map<?, ?> map)
  {
    verify(map, new Options());
  }
  public static void verify(Map<?, ?> map, Options options)
  {
    verify(StringUtils.toString(map), options);
  }
  public static void verify(ResultSet rs)
  {
    verify(rs, new Options());
  }
  public static void verify(ResultSet rs, Options options)
  {
    verify(options.createWriter(rs), options);
  }
  public static <T> void verify(SqlLoader<T> loader)
  {
    verify(loader, new Options());
  }
  public static <T> void verify(SqlLoader<T> loader, Options options)
  {
    verify(new SqlLoader.ExecutableWrapper<T>(loader), options);
  }
  public static ApprovalNamer createApprovalNamer()
  {
    return namerCreater.load();
  }
  public static void verifyEachFileInDirectory(File directory)
  {
    verifyEachFileInDirectory(directory, new Options());
  }
  public static void verifyEachFileInDirectory(File directory, Options options)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(), options);
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter)
  {
    verifyEachFileInDirectory(directory, filter, new Options());
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter, Options options)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter), options);
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter)
  {
    verifyEachFileInDirectory(directory, filter, new Options());
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter, Options options)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter), options);
  }
  private static void verifyEachFileAgainstMasterDirectory(File[] files, Options options)
  {
    List<File> mismatched = new ArrayList<>();
    for (File f : files)
    {
      if (!f.isDirectory())
      {
        try
        {
          verify(options.createWriter(f), new MasterDirectoryNamer(f, options), options);
        }
        catch (Throwable e)
        {
          mismatched.add(f);
        }
      }
    }
    if (!mismatched.isEmpty())
    {
      String message = "The Following Files did not match up: " + getFileNameList(mismatched);
      throw new Error(message);
    }
  }
  private static String getFileNameList(List<File> mismatched)
  {
    return Query.select(mismatched, a -> a.getName()).toString();
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyJson(String json)
  {
    JsonApprovals.verifyJson(json);
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyJson(String json, Options options)
  {
    JsonApprovals.verifyJson(json, options);
  }
  /**
   * @deprecated View method source for example.
   */
  @Deprecated
  public static void verify(Object actual, String fileExtensionWithoutDot)
  {
    verify(actual, fileExtensionWithoutDot, new Options());
  }
  /**
   * @deprecated View method source for example.
   */
  @Deprecated
  public static void verify(Object actual, String fileExtensionWithoutDot, Options options)
  {
    verify("" + actual, options.forFile().withExtension(fileExtensionWithoutDot));
  }
  public static ApprovalFailureReporter getReporter()
  {
    return ReporterFactory.get();
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyAsJson(Object o)
  {
    JsonApprovals.verifyAsJson(o);
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static <T> void verifyAsJson(Object o, Function1<T, T> gsonBuilder, Class<T> gsonBuilderClass)
  {
    verifyAsJson(o, gsonBuilder, gsonBuilderClass, new Options());
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static <T> void verifyAsJson(Object o, Function1<T, T> gsonBuilder, Class<T> gsonBuilderClass,
      Options options)
  {
    verify(JsonUtils.asJsonWithBuilder(o, gsonBuilder, gsonBuilderClass),
        options.forFile().withExtension(".json"));
  }
  /**
   * @deprecated Use JsonApprovals
   */
  @Deprecated
  public static void verifyAsJson(Object o, Options options)
  {
    JsonApprovals.verifyAsJson(o, options);
  }
  public static void verifyException(Action0 runnableBlock)
  {
    verifyException(runnableBlock, new Options());
  }
  public static void verifyException(Action0 runnableBlock, Options options)
  {
    Throwable t = ObjectUtils.captureException(runnableBlock);
    if (t == null)
    { throw new FormattedException("No exception thrown when running %s", runnableBlock); }
    Approvals.verify(String.format("%s: %s", t.getClass().getName(), t.getMessage()), options);
  }
  public static ApprovalSettings settings()
  {
    return new ApprovalSettings();
  }
}
