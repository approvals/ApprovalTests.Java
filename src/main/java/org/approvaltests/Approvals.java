package org.approvaltests;

import java.awt.Component;
import java.awt.Image;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import org.apache.hadoop.fs.Path;
import org.approvaltests.ReporterFactory.FileTypes;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.ApprovalFailureOverrider;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalWriter;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.ApprovalNamerFactory;
import org.approvaltests.namer.JUnitStackTraceNamer;
import org.approvaltests.reporters.ExecutableQueryFailure;
import org.approvaltests.writers.ApprovalBinaryFileWriter;
import org.approvaltests.writers.ApprovalTextWriter;
import org.approvaltests.writers.ApprovalXmlWriter;
import org.approvaltests.writers.ComponentApprovalWriter;
import org.approvaltests.writers.DirectoryToDirectoryWriter;
import org.approvaltests.writers.FileApprovalWriter;
import org.approvaltests.writers.ImageApprovalWriter;
import org.approvaltests.writers.ResultSetApprovalWriter;
import org.jrack.RackResponse;
import org.jrack.RackResponseUtils;
import org.lambda.functions.Function1;
import org.lambda.functions.implementations.F1;
import org.lambda.query.Query;

import com.spun.util.ArrayUtils;
import com.spun.util.ObjectUtils;
import com.spun.util.StringUtils;
import com.spun.util.images.ImageWriter;
import com.spun.util.persistence.ExecutableQuery;
import com.spun.util.persistence.SqlLoader;

public class Approvals
{
  public static void verify(String response) throws Exception
  {
    verify(new ApprovalTextWriter(response, "txt"), FileTypes.Text);
  }
  public static void verify(Object o) throws Exception
  {
    verify("" + o);
  }
  public static <T> void verifyAll(String label, T[] array) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"), FileTypes.Text);
  }
  public static <T> void verifyAll(String header, String label, T[] array) throws Exception
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"),
        FileTypes.Text);
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1)
  {
    String text = ArrayUtils.toString(values, f1);
    verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, Arrays.asList(values), f1);
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1)
  {
    String text = formatHeader(header) + ArrayUtils.toString(array, f1);
    verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
  }
  private static String formatHeader(String header)
  {
    return StringUtils.isEmpty(header) ? "" : header + "\r\n\r\n\r\n";
  }
  public static <T> void verifyAll(String label, Iterable<T> array) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"), FileTypes.Text);
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array) throws Exception
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"),
        FileTypes.Text);
  }
  public static void verify(Component c) throws Exception
  {
    verify(new ComponentApprovalWriter(c), FileTypes.Image);
  }
  public static void verifyHtml(String response) throws Exception
  {
    verify(new ApprovalTextWriter(response, "html"), FileTypes.Html);
  }
  public static void verify(File generateFile)
  {
    verify(new FileApprovalWriter(generateFile), FileTypes.File);
  }
  public static void verify(Image image)
  {
    approve(ImageWriter.toBufferedImage(image), createApprovalNamer());
  }
  public static void verify(BufferedImage bufferedImage)
  {
    verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
  }
  public static void verify(ApprovalWriter writter, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(new FileApprover(writter, namer), reporter);
  }
  public static void verify(ApprovalWriter writter, String fileType)
  {
    verify(writter, createApprovalNamer(), ReporterFactory.get(fileType));
  }
  public static void verifyXml(String xml) throws Exception
  {
    verify(new ApprovalXmlWriter(xml), FileTypes.Text);
  }
  public static void verify(FileApprover approver, ApprovalFailureReporter reporter)
  {
    try
    {
      if (!approver.approve())
      {
        boolean passed = false;
        if (reporter instanceof ApprovalFailureOverrider)
        {
          passed = approver.askToChangeReceivedToApproved((ApprovalFailureOverrider) reporter);
        }
        if (!passed)
        {
          approver.reportFailure(reporter);
          approver.fail();
        }
        else
        {
          approver.cleanUpAfterSuccess(reporter);
        }
      }
      else
      {
        approver.cleanUpAfterSuccess(reporter);
      }
    }
    catch (Exception e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
  public static void verify(ExecutableQuery query) throws Exception
  {
    verify(new ApprovalTextWriter(query.getQuery(), "txt"), createApprovalNamer(), new ExecutableQueryFailure(
        query));
  }
  public static void verify(Map map) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(map), "txt"), FileTypes.Text);
  }
  public static void verify(RackResponse response) throws Exception
  {
    if (isImage(response))
    {
      String fileType = "png";
      verify(new ApprovalBinaryFileWriter(response.getResponse(), fileType), fileType);
    }
    else
    {
      verifyHtml(response.getResponse().toString());
    }
  }
  public static void verify(ResultSet rs) throws Exception
  {
    verify(new ResultSetApprovalWriter(rs), "csv");
  }
  public static void verify(SqlLoader loader) throws Exception
  {
    verify(new SqlLoader.ExecutableWrapper(loader));
  }
  private static boolean isImage(RackResponse response)
  {
    String type = response.getHeaders().get(RackResponseUtils.CONTENT_TYPE);
    return RackResponseUtils.CONTENT_TYPE_IMAGE.equals(type);
  }
  public static ApprovalNamer createApprovalNamer()
  {
    return ApprovalNamerFactory.getApprovalNamer();
  }
  private static void approve(BufferedImage bufferedImage, ApprovalNamer namer)
  {
    verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
  }
  public static void verifyEachFileInDirectory(File directory)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles());
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  private static void verifyEachFileAgainstMasterDirectory(File[] files) throws Error
  {
    ApprovalNamer namer = createApprovalNamer();
    String dirName = namer.getApprovalFileBasePath() + Path.SEPARATOR + namer.getApprovalName() + ".Files";
    File approvedDirectory = new File(dirName);
    List<File> mismatched = new ArrayList<File>();
    for (File f : files)
    {
      if (!f.isDirectory())
      {
        try
        {
          verify(new DirectoryToDirectoryWriter(f, approvedDirectory), FileTypes.File);
        }
        catch (Throwable e)
        {
          mismatched.add(f);
        }
      }
    }
    if (!mismatched.isEmpty())
    {
      String message = "The Following Files did not match up: " + getFileNameList(mismatched);
      throw new Error(message);
    }
  }
  private static String getFileNameList(List<File> mismatched)
  {
    return Query.select(mismatched, new F1<File, String>(mismatched.get(1))
    {
      {
        ret(a.getName());
      }
    }).toString();
  }
}