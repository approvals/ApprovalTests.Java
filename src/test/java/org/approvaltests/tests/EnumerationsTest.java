package org.approvaltests.tests;

import junit.framework.TestCase;

import org.approvaltests.Approvals;
import org.approvaltests.strings.Printer;

public class EnumerationsTest extends TestCase
{
  public void testNumbers() throws Exception
  {
    Approvals.verifyAll("i", new Integer[]{5, 4, 3, 2, 1});
  }
  public void testNumbersWithHeader() throws Exception
  {
    Approvals.verifyAll("The Numbers", "i", new Integer[]{5, 4, 3, 2, 1});
  }
  public void testNumbersWithLambdas() throws Exception
  {
    String[] numbers = new String[]{"one", "two", "three", "four"};
    Approvals.verifyAll(numbers, new Printer<String>("")
    {
      {
        print(a, a.length());
      }
    });
  }
  public void testNumbersWithLambdasAndHeader() throws Exception
  {
    String[] numbers = new String[]{"one", "two", "three", "four"};
    Approvals.verifyAll("Lengths of Strings", numbers, new Printer<String>("")
    {
      {
        print(a, a.length());
      }
    });
  }
}
