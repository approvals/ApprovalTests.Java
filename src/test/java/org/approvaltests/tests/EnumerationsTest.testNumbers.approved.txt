i[0] = 5
i[1] = 4
i[2] = 3
i[3] = 2
i[4] = 1
